File: backend/ai-detection/lib/circuit_breaker.py
"""
Circuit Breaker Pattern Implementation
Prevents cascading failures when external services are unavailable
"""
import asyncio
import logging
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable, Dict, Optional, Any, TypeVar, Generic
from functools import wraps
import threading

logger = logging.getLogger(__name__)

T = TypeVar('T')


class CircuitState(Enum):
    """Circuit breaker states"""
    CLOSED = "closed"      # Normal operation, requests flow through
    OPEN = "open"          # Failing, requests are blocked
    HALF_OPEN = "half_open"  # Testing if service recovered


@dataclass
class CircuitStats:
    """Statistics for a circuit breaker"""
    total_calls: int = 0
    successful_calls: int = 0
    failed_calls: int = 0
    rejected_calls: int = 0
    last_failure_time: Optional[float] = None
    last_success_time: Optional[float] = None
    consecutive_failures: int = 0
    consecutive_successes: int = 0


@dataclass
class CircuitBreakerConfig:
    """Configuration for a circuit breaker"""
    # Failure threshold - number of failures before opening circuit
    failure_threshold: int = 5
    # Failure rate threshold (0.0 to 1.0) - alternative to count
    failure_rate_threshold: float = 0.5
    # Minimum number of calls before failure rate is evaluated
    minimum_calls: int = 10
    # Time in seconds to wait before attempting recovery (half-open)
    recovery_timeout: float = 30.0
    # Number of successful calls in half-open state to close circuit
    success_threshold: int = 3
    # Time window in seconds for failure rate calculation
    failure_window: float = 60.0
    # Optional fallback function
    fallback: Optional[Callable] = None


class CircuitBreaker(Generic[T]):
    """
    Circuit Breaker implementation for protecting external service calls.

    States:
    - CLOSED: Normal operation, requests pass through
    - OPEN: Service is failing, requests are blocked (fail fast)
    - HALF_OPEN: Testing if service recovered

    Example usage:
        breaker = CircuitBreaker(
            name="openai",
            config=CircuitBreakerConfig(failure_threshold=5)
        )

        @breaker
        async def call_openai(prompt: str) -> str:
            return await openai_client.complete(prompt)
    """

    def __init__(self, name: str, config: Optional[CircuitBreakerConfig] = None):
        self.name = name
        self.config = config or CircuitBreakerConfig()
        self._state = CircuitState.CLOSED
        self._stats = CircuitStats()
        self._last_state_change = time.time()
        self._lock = threading.RLock()
        self._failure_times: list[float] = []

    @property
    def state(self) -> CircuitState:
        """Get current circuit state, checking for automatic state transitions"""
        with self._lock:
            if self._state == CircuitState.OPEN:
                # Check if recovery timeout has elapsed
                if time.time() - self._last_state_change >= self.config.recovery_timeout:
                    self._transition_to(CircuitState.HALF_OPEN)
            return self._state

    @property
    def stats(self) -> CircuitStats:
        """Get circuit statistics"""
        return self._stats

    @property
    def is_closed(self) -> bool:
        """Check if circuit is closed (allowing requests)"""
        return self.state == CircuitState.CLOSED

    @property
    def is_open(self) -> bool:
        """Check if circuit is open (blocking requests)"""
        return self.state == CircuitState.OPEN

    @property
    def is_half_open(self) -> bool:
        """Check if circuit is in half-open state (testing)"""
        return self.state == CircuitState.HALF_OPEN

    def _transition_to(self, new_state: CircuitState) -> None:
        """Transition to a new state"""
        old_state = self._state
        self._state = new_state
        self._last_state_change = time.time()

        logger.info(
            f"Circuit breaker '{self.name}' transitioned from {old_state.value} to {new_state.value}"
        )

        # Reset counters on state change
        if new_state == CircuitState.CLOSED:
            self._stats.consecutive_failures = 0
            self._stats.consecutive_successes = 0
        elif new_state == CircuitState.HALF_OPEN:
            self._stats.consecutive_successes = 0

    def _record_success(self) -> None:
        """Record a successful call"""
        with self._lock:
            self._stats.total_calls += 1
            self._stats.successful_calls += 1
            self._stats.last_success_time = time.time()
            self._stats.consecutive_successes += 1
            self._stats.consecutive_failures = 0

            # Check for state transition
            if self._state == CircuitState.HALF_OPEN:
                if self._stats.consecutive_successes >= self.config.success_threshold:
                    self._transition_to(CircuitState.CLOSED)

    def _record_failure(self) -> None:
        """Record a failed call"""
        with self._lock:
            current_time = time.time()
            self._stats.total_calls += 1
            self._stats.failed_calls += 1
            self._stats.last_failure_time = current_time
            self._stats.consecutive_failures += 1
            self._stats.consecutive_successes = 0

            # Track failure times for rate calculation
            self._failure_times.append(current_time)
            # Remove old failures outside the window
            cutoff = current_time - self.config.failure_window
            self._failure_times = [t for t in self._failure_times if t > cutoff]

            # Check for state transition
            if self._state == CircuitState.HALF_OPEN:
                # Single failure in half-open opens the circuit
                self._transition_to(CircuitState.OPEN)
            elif self._state == CircuitState.CLOSED:
                # Check failure threshold
                should_open = False

                # Check consecutive failure count
                if self._stats.consecutive_failures >= self.config.failure_threshold:
                    should_open = True
                    logger.warning(
                        f"Circuit breaker '{self.name}': consecutive failures "
                        f"({self._stats.consecutive_failures}) exceeded threshold "
                        f"({self.config.failure_threshold})"
                    )

                # Check failure rate
                if (self._stats.total_calls >= self.config.minimum_calls and
                    len(self._failure_times) >= self.config.minimum_calls):
                    failure_rate = len(self._failure_times) / self._stats.total_calls
                    if failure_rate >= self.config.failure_rate_threshold:
                        should_open = True
                        logger.warning(
                            f"Circuit breaker '{self.name}': failure rate "
                            f"({failure_rate:.2%}) exceeded threshold "
                            f"({self.config.failure_rate_threshold:.2%})"
                        )

                if should_open:
                    self._transition_to(CircuitState.OPEN)

    def _record_rejected(self) -> None:
        """Record a rejected call (circuit open)"""
        with self._lock:
            self._stats.rejected_calls += 1

    async def call(
        self,
        func: Callable[..., T],
        *args,
        fallback: Optional[Callable[..., T]] = None,
        **kwargs
    ) -> T:
        """
        Execute a function through the circuit breaker.

        Args:
            func: Async function to execute
            *args: Positional arguments for the function
            fallback: Optional fallback function if circuit is open
            **kwargs: Keyword arguments for the function

        Returns:
            Result of the function or fallback

        Raises:
            CircuitOpenError: If circuit is open and no fallback is provided
        """
        state = self.state

        if state == CircuitState.OPEN:
            self._record_rejected()
            fallback_fn = fallback or self.config.fallback
            if fallback_fn:
                logger.debug(f"Circuit breaker '{self.name}' is open, using fallback")
                if asyncio.iscoroutinefunction(fallback_fn):
                    return await fallback_fn(*args, **kwargs)
                return fallback_fn(*args, **kwargs)
            raise CircuitOpenError(
                f"Circuit breaker '{self.name}' is open. "
                f"Last failure: {self._stats.last_failure_time}, "
                f"Consecutive failures: {self._stats.consecutive_failures}"
            )

        try:
            if asyncio.iscoroutinefunction(func):
                result = await func(*args, **kwargs)
            else:
                result = func(*args, **kwargs)
            self._record_success()
            return result
        except Exception as e:
            self._record_failure()
            raise

    def __call__(self, func: Callable) -> Callable:
        """
        Decorator to wrap a function with circuit breaker protection.

        Example:
            @circuit_breaker
            async def call_external_api():
                return await api.get_data()
        """
        @wraps(func)
        async def async_wrapper(*args, **kwargs):
            return await self.call(func, *args, **kwargs)

        @wraps(func)
        def sync_wrapper(*args, **kwargs):
            return asyncio.get_event_loop().run_until_complete(
                self.call(func, *args, **kwargs)
            )

        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper

    def reset(self) -> None:
        """Reset the circuit breaker to closed state"""
        with self._lock:
            self._state = CircuitState.CLOSED
            self._stats = CircuitStats()
            self._failure_times = []
            self._last_state_change = time.time()
            logger.info(f"Circuit breaker '{self.name}' reset to closed state")

    def to_dict(self) -> Dict[str, Any]:
        """Get circuit breaker status as dictionary"""
        return {
            "name": self.name,
            "state": self.state.value,
            "stats": {
                "total_calls": self._stats.total_calls,
                "successful_calls": self._stats.successful_calls,
                "failed_calls": self._stats.failed_calls,
                "rejected_calls": self._stats.rejected_calls,
                "consecutive_failures": self._stats.consecutive_failures,
                "consecutive_successes": self._stats.consecutive_successes,
                "last_failure_time": self._stats.last_failure_time,
                "last_success_time": self._stats.last_success_time,
            },
            "config": {
                "failure_threshold": self.config.failure_threshold,
                "failure_rate_threshold": self.config.failure_rate_threshold,
                "recovery_timeout": self.config.recovery_timeout,
                "success_threshold": self.config.success_threshold,
            }
        }


class CircuitOpenError(Exception):
    """Exception raised when circuit breaker is open"""
    pass


class CircuitBreakerRegistry:
    """
    Registry for managing multiple circuit breakers.
    Provides centralized access and monitoring.
    """

    _instance: Optional['CircuitBreakerRegistry'] = None
    _lock = threading.Lock()

    def __new__(cls) -> 'CircuitBreakerRegistry':
        with cls._lock:
            if cls._instance is None:
                cls._instance = super().__new__(cls)
                cls._instance._breakers: Dict[str, CircuitBreaker] = {}
            return cls._instance

    def register(self, breaker: CircuitBreaker) -> None:
        """Register a circuit breaker"""
        self._breakers[breaker.name] = breaker
        logger.debug(f"Registered circuit breaker: {breaker.name}")

    def get(self, name: str) -> Optional[CircuitBreaker]:
        """Get a circuit breaker by name"""
        return self._breakers.get(name)

    def get_or_create(
        self,
        name: str,
        config: Optional[CircuitBreakerConfig] = None
    ) -> CircuitBreaker:
        """Get an existing circuit breaker or create a new one"""
        if name not in self._breakers:
            breaker = CircuitBreaker(name, config)
            self.register(breaker)
        return self._breakers[name]

    def get_all_status(self) -> Dict[str, Dict[str, Any]]:
        """Get status of all circuit breakers"""
        return {
            name: breaker.to_dict()
            for name, breaker in self._breakers.items()
        }

    def get_open_circuits(self) -> Dict[str, CircuitBreaker]:
        """Get all open circuit breakers"""
        return {
            name: breaker
            for name, breaker in self._breakers.items()
            if breaker.is_open
        }

    def reset_all(self) -> None:
        """Reset all circuit breakers"""
        for breaker in self._breakers.values():
            breaker.reset()


# Global registry instance
def get_circuit_registry() -> CircuitBreakerRegistry:
    """Get the global circuit breaker registry"""
    return CircuitBreakerRegistry()


# Convenience function to create a circuit breaker
def create_circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 30.0,
    failure_rate_threshold: float = 0.5,
    **kwargs
) -> CircuitBreaker:
    """
    Create and register a circuit breaker.

    Args:
        name: Unique name for the circuit breaker
        failure_threshold: Number of consecutive failures to open circuit
        recovery_timeout: Seconds to wait before attempting recovery
        failure_rate_threshold: Failure rate (0.0-1.0) to trigger open
        **kwargs: Additional config options

    Returns:
        Configured CircuitBreaker instance
    """
    config = CircuitBreakerConfig(
        failure_threshold=failure_threshold,
        recovery_timeout=recovery_timeout,
        failure_rate_threshold=failure_rate_threshold,
        **kwargs
    )
    breaker = CircuitBreaker(name, config)
    get_circuit_registry().register(breaker)
    return breaker

File: backend/api-gateway/lib/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 * Provides fault tolerance for external service calls
 *
 * States:
 * - CLOSED: Normal operation, requests flow through
 * - OPEN: Circuit tripped, requests fail fast
 * - HALF_OPEN: Testing if service has recovered
 */

import { ServiceUnavailableError } from './errors';

export enum CircuitState {
  CLOSED = 'CLOSED',
  OPEN = 'OPEN',
  HALF_OPEN = 'HALF_OPEN',
}

export interface CircuitBreakerOptions {
  /** Name for logging/identification */
  name: string;
  /** Number of failures before opening circuit (default: 5) */
  failureThreshold?: number;
  /** Number of successful calls to close circuit (default: 3) */
  successThreshold?: number;
  /** Time in ms before attempting recovery (default: 30000) */
  timeout?: number;
  /** Time window in ms to count failures (default: 60000) */
  failureWindow?: number;
  /** Custom function to determine if error should count as failure */
  isFailure?: (error: Error) => boolean;
  /** Callback when circuit state changes */
  onStateChange?: (from: CircuitState, to: CircuitState, name: string) => void;
}

export interface CircuitBreakerStats {
  name: string;
  state: CircuitState;
  failures: number;
  successes: number;
  lastFailureTime: number | null;
  lastSuccessTime: number | null;
  totalRequests: number;
  totalFailures: number;
  totalSuccesses: number;
}

/**
 * Circuit Breaker Implementation
 */
export class CircuitBreaker {
  private name: string;
  private state: CircuitState = CircuitState.CLOSED;
  private failureThreshold: number;
  private successThreshold: number;
  private timeout: number;
  private failureWindow: number;
  private isFailure: (error: Error) => boolean;
  private onStateChange?: (from: CircuitState, to: CircuitState, name: string) => void;

  // Tracking
  private failures: number = 0;
  private successes: number = 0;
  private lastFailureTime: number | null = null;
  private lastSuccessTime: number | null = null;
  private openedAt: number | null = null;
  private failureTimestamps: number[] = [];

  // Stats
  private totalRequests: number = 0;
  private totalFailures: number = 0;
  private totalSuccesses: number = 0;

  constructor(options: CircuitBreakerOptions) {
    this.name = options.name;
    this.failureThreshold = options.failureThreshold ?? 5;
    this.successThreshold = options.successThreshold ?? 3;
    this.timeout = options.timeout ?? 30000; // 30 seconds
    this.failureWindow = options.failureWindow ?? 60000; // 60 seconds
    this.isFailure = options.isFailure ?? (() => true);
    this.onStateChange = options.onStateChange;
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    this.totalRequests++;

    // Check if circuit should transition from OPEN to HALF_OPEN
    if (this.state === CircuitState.OPEN) {
      if (this.shouldAttemptReset()) {
        this.transitionTo(CircuitState.HALF_OPEN);
      } else {
        throw new ServiceUnavailableError(
          `Circuit breaker is open for service: ${this.name}`
        );
      }
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure(error as Error);
      throw error;
    }
  }

  /**
   * Check if we should attempt to reset the circuit
   */
  private shouldAttemptReset(): boolean {
    if (!this.openedAt) return false;
    return Date.now() - this.openedAt >= this.timeout;
  }

  /**
   * Handle successful execution
   */
  private onSuccess(): void {
    this.lastSuccessTime = Date.now();
    this.totalSuccesses++;

    if (this.state === CircuitState.HALF_OPEN) {
      this.successes++;
      if (this.successes >= this.successThreshold) {
        this.transitionTo(CircuitState.CLOSED);
      }
    } else if (this.state === CircuitState.CLOSED) {
      // Reset failure count on success in closed state
      this.failures = 0;
      this.failureTimestamps = [];
    }
  }

  /**
   * Handle failed execution
   */
  private onFailure(error: Error): void {
    // Check if this error should count as a failure
    if (!this.isFailure(error)) {
      return;
    }

    this.lastFailureTime = Date.now();
    this.totalFailures++;

    if (this.state === CircuitState.HALF_OPEN) {
      // Any failure in half-open state reopens the circuit
      this.transitionTo(CircuitState.OPEN);
    } else if (this.state === CircuitState.CLOSED) {
      this.recordFailure();

      // Count failures within the window
      const recentFailures = this.getRecentFailures();
      if (recentFailures >= this.failureThreshold) {
        this.transitionTo(CircuitState.OPEN);
      }
    }
  }

  /**
   * Record a failure timestamp
   */
  private recordFailure(): void {
    const now = Date.now();
    this.failureTimestamps.push(now);
    this.failures++;

    // Clean up old timestamps
    const windowStart = now - this.failureWindow;
    this.failureTimestamps = this.failureTimestamps.filter(t => t > windowStart);
  }

  /**
   * Get count of failures within the window
   */
  private getRecentFailures(): number {
    const windowStart = Date.now() - this.failureWindow;
    return this.failureTimestamps.filter(t => t > windowStart).length;
  }

  /**
   * Transition to a new state
   */
  private transitionTo(newState: CircuitState): void {
    const oldState = this.state;
    this.state = newState;

    if (newState === CircuitState.OPEN) {
      this.openedAt = Date.now();
      this.successes = 0;
    } else if (newState === CircuitState.CLOSED) {
      this.failures = 0;
      this.successes = 0;
      this.failureTimestamps = [];
      this.openedAt = null;
    } else if (newState === CircuitState.HALF_OPEN) {
      this.successes = 0;
    }

    if (this.onStateChange && oldState !== newState) {
      this.onStateChange(oldState, newState, this.name);
    }
  }

  /**
   * Get current state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Get statistics
   */
  getStats(): CircuitBreakerStats {
    return {
      name: this.name,
      state: this.state,
      failures: this.failures,
      successes: this.successes,
      lastFailureTime: this.lastFailureTime,
      lastSuccessTime: this.lastSuccessTime,
      totalRequests: this.totalRequests,
      totalFailures: this.totalFailures,
      totalSuccesses: this.totalSuccesses,
    };
  }

  /**
   * Manually reset the circuit breaker
   */
  reset(): void {
    this.transitionTo(CircuitState.CLOSED);
  }

  /**
   * Manually open the circuit breaker
   */
  trip(): void {
    this.transitionTo(CircuitState.OPEN);
  }
}

// ============================================================================
// Circuit Breaker Registry
// ============================================================================

const circuitBreakers = new Map<string, CircuitBreaker>();

/**
 * Get or create a circuit breaker for a service
 */
export function getCircuitBreaker(options: CircuitBreakerOptions): CircuitBreaker {
  const existing = circuitBreakers.get(options.name);
  if (existing) {
    return existing;
  }

  const breaker = new CircuitBreaker(options);
  circuitBreakers.set(options.name, breaker);
  return breaker;
}

/**
 * Get all circuit breaker statistics
 */
export function getAllCircuitBreakerStats(): CircuitBreakerStats[] {
  return Array.from(circuitBreakers.values()).map(cb => cb.getStats());
}

/**
 * Reset all circuit breakers
 */
export function resetAllCircuitBreakers(): void {
  circuitBreakers.forEach(cb => cb.reset());
}

// ============================================================================
// Pre-configured Circuit Breakers for Services
// ============================================================================

const defaultStateChangeHandler = (from: CircuitState, to: CircuitState, name: string) => {
  console.log(`[CircuitBreaker] ${name}: ${from} -> ${to}`);
};

export function getAuthServiceCircuitBreaker(): CircuitBreaker {
  return getCircuitBreaker({
    name: 'auth-service',
    failureThreshold: 5,
    successThreshold: 3,
    timeout: 30000,
    failureWindow: 60000,
    onStateChange: defaultStateChangeHandler,
  });
}

export function getSessionServiceCircuitBreaker(): CircuitBreaker {
  return getCircuitBreaker({
    name: 'session-service',
    failureThreshold: 5,
    successThreshold: 3,
    timeout: 30000,
    failureWindow: 60000,
    onStateChange: defaultStateChangeHandler,
  });
}

export function getAiDetectionServiceCircuitBreaker(): CircuitBreaker {
  return getCircuitBreaker({
    name: 'ai-detection-service',
    failureThreshold: 3, // More sensitive for AI service
    successThreshold: 2,
    timeout: 60000, // Longer timeout for AI service
    failureWindow: 120000,
    onStateChange: defaultStateChangeHandler,
  });
}

export function getEyeTrackingServiceCircuitBreaker(): CircuitBreaker {
  return getCircuitBreaker({
    name: 'eye-tracking-service',
    failureThreshold: 5,
    successThreshold: 3,
    timeout: 30000,
    failureWindow: 60000,
    onStateChange: defaultStateChangeHandler,
  });
}

// ============================================================================
// Utility: Wrap HTTP Client with Circuit Breaker
// ============================================================================

/**
 * Execute an HTTP request with circuit breaker protection
 */
export async function withCircuitBreaker<T>(
  circuitBreaker: CircuitBreaker,
  fn: () => Promise<T>
): Promise<T> {
  return circuitBreaker.execute(fn);
}
